File: Prototype/BigGp/src/Model/BaseNodes/ForStatement/ForIncrement.java
package Model.BaseNodes.ForStatement;

import Model.Interfaces.Crossable;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Constants.AddRandomConstant;
import Model.SmallNodes.Constants.Constant;
import Model.SmallNodes.Expressions.Standart.Childs.VarExpression;
import Model.SmallNodes.Operators.Arithmetic.AdditiveOperator;
import Serialization.Token;

import java.util.ArrayList;
import java.util.Random;

public class ForIncrement extends Node implements Crossable, SubtreeMutable {

    private AdditiveOperator additiveOperator;
    private VarExpression variable;
    private Constant constInt;

    @Override
    public String toString() {

        if (variable != null) {
            return additiveOperator.getOperator() + variable.toString();
        } else {
            return additiveOperator.getOperator() + constInt.toString();
        }
    }

    public ArrayList<Node> getChildrenAsNodes() {
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);
        nodes.addAll(additiveOperator.getChildrenAsNodes());

        if (variable != null) {
            nodes.addAll(variable.getChildrenAsNodes());
        } else {
            nodes.addAll(constInt.getChildrenAsNodes());
        }

        return nodes;
    }

    @Override
    public void subtreeMutate() {
        if (this.getParentNode() instanceof ForStatement forStatement){
            forStatement.setForIncrement(new ForIncrement(forStatement));
        }
    }

    public ForIncrement(Node parentNode) {
        super(parentNode, "ForIncrement", true);

        variable = null;
        constInt = null;

        if (new Random().nextInt(2) == 0) {
            variable = new VarExpression(this, true);
        } else {
            constInt = new Constant(this, AddRandomConstant.RandomConstant.INT);
        }
        additiveOperator = new AdditiveOperator(this);
    }

    private void forIncrementFromTokens(ArrayList<Token> tokens) {
        switch (tokens.get(0).getNAME()) {
            case "VarExpression" -> variable = new VarExpression(this, tokens);
            case "AdditiveOperator" -> additiveOperator = new AdditiveOperator(this, tokens);
            case "Constant" -> constInt = new Constant(this, AddRandomConstant.RandomConstant.INT, tokens);
        }
    }

    public ForIncrement(Node parent, ArrayList<Token> tokens) {
        super(parent, "ForIncrement", true);
        tokens.remove(0);

        forIncrementFromTokens(tokens);
        forIncrementFromTokens(tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Big/For/ForIncrement.cs
using Model.Abstract;
using Model.Interfaces;
using Model.Nodes.Small.Constants;
using Model.Nodes.Small.Expressions.Standard;
using Model.Nodes.Small.Operators.Arithmetic;
using Utils;

namespace Model.Nodes.Big.For;

public sealed class ForIncrement : Node, ICrossable, ISubtreeMutable
{
    private AdditiveOperator _additiveOperator;
    private VarExpression? _variable;
    private Constant? _constant;

    public override List<Node> ChildrenAsNodes()
    {
        var nodes = new List<Node> { this };
        nodes.AddRange(_additiveOperator.ChildrenAsNodes());
        nodes.AddRange(_variable is null
            ? _constant?.ChildrenAsNodes() ?? throw new InvalidOperationException("Both variable and constant are null")
            : _variable.ChildrenAsNodes()
        );
        
        return nodes;
    }

    public void SubtreeMutate()
    {
        if (ParentNode is ForStatement forStatement)
        {
            forStatement.ForIncrement = new ForIncrement(forStatement);
        }
    }

    private void FromTokens(IList<Token> tokens)
    {
        switch (tokens[0].Name)
        {
            case "VarExpression":
                _variable = new VarExpression(this, tokens);
                break;
            case "AdditiveOperator":
                _additiveOperator = new AdditiveOperator(this, tokens);
                break;
            case "Constant":
                _constant = new Constant(this, ConstantType.Int, tokens);
                break;
            default:
                throw new InvalidOperationException($"Unknown token: {tokens[0]}");
        }
    }
    
    public ForIncrement(Node parentNode) : base(parentNode, "ForIncrement", true)
    {
        if (RandomService.RandomPercentage() < 0.50)
        {
            _variable = new VarExpression(this, true);
        }
        else
        {
            _constant = new Constant(this, ConstantType.Int);
        }
        _additiveOperator = new AdditiveOperator(this);
    }
    
    public ForIncrement(Node parentNode, IList<Token> tokens) : base(parentNode, "ForIncrement", true)
    {
        tokens.RemoveAt(0);
        FromTokens(tokens);
        FromTokens(tokens);
    }

    public override string ToString()
        => _variable is null
            ? $"{_additiveOperator.OperatorValue}{_constant}"
            : $"{_additiveOperator.OperatorValue}{_variable}";
}